src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,331 @@
+      if (filter->get_d3n_cache()->get(oid_in_cache, obj_max_req_size)) {
+        // Read From Cache
+        auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+        lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+        auto r = flush(dpp, std::move(completed));
+        if (r < 0) {
+          lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          return r;
+        }
+      } else {
+        lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+        auto r = drain(dpp);
+        if (r < 0) {
+          lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          return r;
+        }
+        break;
+      }
+    }
+    if (start_part_num == (num_parts - 1)) {
+      lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+      return drain(dpp);
+    } else {
+      adjusted_start_ofs += obj_max_req_size;
+    }
+    start_part_num += 1;
+    len -= obj_max_req_size;
+  } while(start_part_num < num_parts);
+  lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+  Attrs obj_attrs;
+  if (source->has_attrs()) {
+    obj_attrs = source->get_attrs();
+  }
+  const bool is_compressed = (obj_attrs.find(RGW_ATTR_COMPRESSION) != obj_attrs.end());
+  const bool is_encrypted = (obj_attrs.find(RGW_ATTR_CRYPT_MODE) != obj_attrs.end());
+  lderr(dpp->get_cct()) << "D3NFilterObject::iterate:: " << __func__ << "is_compressed: " << is_compressed << " is_encrypted: " << is_encrypted << dendl;
+  if (source->is_compressed() || obj_attrs.find(RGW_ATTR_CRYPT_MODE) != obj_attrs.end()) {
```

Comment:
```suggestion
  if (is_compressed || is_encrypted) {
```

Reply:
I dont think read_ofs != 0 is needed here, and I am not sure what astate->size != astate->accounted_size. d3n filter driver has a chunking mechanism different from that of the rados store (which is used by legacy d3n also).